src/scheduler.rs

```diff
@@ -0,0 +1,1041 @@
+    }
+}
+
+impl<'a> GroupAntiAffinityStrategy<'a> {
+    pub fn new(eligible_nodes: RoleGroupEligibleNodes, pod_node_map: &'a PodToNodeMapping) -> Self {
+        GroupAntiAffinityStrategy {
+            eligible_nodes: RefCell::new(eligible_nodes),
+            existing_mapping: pod_node_map,
+        }
+    }
+
+    pub fn select_node_for_pod(
+        &self,
+        pod_id: &PodIdentity,
+        preferred_node: Option<&NodeIdentity>,
+    ) -> Option<NodeIdentity> {
+        let mut borrowed_nodes = self.eligible_nodes.borrow_mut();
+
+        // Find a node to schedule on (it might be the node from history)
+        while let Some(next_node) = borrowed_nodes.preferred_node_or_last(pod_id, preferred_node) {
+            borrowed_nodes.remove_eligible_node(
+                &next_node,
+                pod_id.role.as_str(),
+                pod_id.group.as_str(),
+            );
+            // check that the node is not already in use
+            if self.existing_mapping.mapped_by(&next_node).is_none() {
+                return Some(next_node);
+            }
+        }
+        None
+    }
+}
+
+impl PodPlacementStrategy for GroupAntiAffinityStrategy<'_> {
+    /// Returns a list of nodes to place to provided pods.
+    /// *Note* Do not call this more than once! This modifies the internal state of the value that
+    /// might not reflect the reality between calls.
+    fn place(
+        &self,
+        pods: &[&PodIdentity],
+        preferred_nodes: &[Option<&NodeIdentity>],
+    ) -> Vec<Option<NodeIdentity>> {
+        assert_eq!(pods.len(), preferred_nodes.len());
+        pods.iter()
+            .zip(preferred_nodes.iter())
+            .map(|(pod, preferred_node)| self.select_node_for_pod(*pod, *preferred_node))
+            .collect()
+    }
+}
+
+impl PodIdentity {
+    pub fn new(app: &str, instance: &str, role: &str, group: &str, id: &str) -> Self {
+        Self::warn_semicolon(app, instance, role, group, id);
+        PodIdentity {
+            app: app.to_string(),
+            instance: instance.to_string(),
+            role: role.to_string(),
+            group: group.to_string(),
+            id: id.to_string(),
+        }
+    }
+
+    pub fn compute_hash(&self, hasher: &mut DefaultHasher) -> u64 {
+        self.hash(hasher);
+        hasher.finish()
+    }
+
+    fn warn_semicolon(app: &str, instance: &str, role: &str, group: &str, id: &str) {
+        if app.contains(SEMICOLON) {
+            warn!("Found semicolon in application name: {}", app);
+        }
+        if instance.contains(SEMICOLON) {
+            warn!("Found semicolon in instance name: {}", instance);
+        }
+        if role.contains(SEMICOLON) {
+            warn!("Found semicolon in role name: {}", role);
+        }
+        if group.contains(SEMICOLON) {
+            warn!("Found semicolon in group name: {}", group);
+        }
+        if id.contains(SEMICOLON) {
+            warn!("Found semicolon in pod id: {}", id);
+        }
```

Comment:
Maybe use `warn!("Found ['{}'] in application name: {}", SEMICOLON, app);` in case it changes?